### PR TITLE
populate uris in eo3 ds with stac url

### PR DESCRIPTION
**Pull request by emmaai**
As the title. Make a `eo3` dataset with `location/uris` populated by stac url, which could to be loaded with `dc.load`.

### CODE REVIEW

**Review by SpacemanPaul — COMMENTED** [review]
The code changes look sensible enough to me, but I would like to see test coverage.

**Comment by emmaai** [comment]
> The code changes look sensible enough to me, but I would like to see test coverage.

The coverage hang there 
https://github.com/opendatacube/odc-stac/actions/runs/8641742492/job/23691767514#step:7:28

File: odc/stac/eo3/_eo3converter.py [hunk 1]
```diff
@@ -313,7 +321,12 @@ def stac2ds(
 
     """
     products: Dict[str, DatasetType] = {} if product_cache is None else product_cache
-    for item in items:
+    if asset_urls is None:
+        asset_urls = []
+
+    items_with_urls = itertools.zip_longest(items, asset_urls)
+
+    for item, url in items_with_urls:
```
Comment (Ariana-B) [inline odc/stac/eo3/_eo3converter.py]: Anything being passed `item` will need to account for the possibility of it being `None`.

File: odc/stac/eo3/_eo3converter.py [hunk 2]
```diff
@@ -245,13 +252,14 @@ def _item_to_ds(
...
 
-    return _to_dataset(_item, item.properties, ds_uuid, product)
+    return _to_dataset(_item, item.properties, ds_uuid, product, asset_url)
 
 
 def stac2ds(
     items: Iterable[pystac.item.Item],
     cfg: Optional[ConversionConfig] = None,
     product_cache: Optional[Dict[str, DatasetType]] = None,
+    asset_urls: Optional[Iterable[str]] = None,
```
Comment (Ariana-B) [inline odc/stac/eo3/_eo3converter.py]: Given that a pystac Item has `assets` with their own hrefs, I think this param might be better called something a little less potentially ambiguous, such as `item_urls`.

File: odc/stac/eo3/_eo3converter.py [hunk 3]
```diff
@@ -210,24 +212,29 @@ def _to_dataset(
     if crs is None:
         crs = EPSG4326
 
+    if asset_url is None:
```
Comment (Ariana-B) [inline odc/stac/eo3/_eo3converter.py]: Would it be worth accounting for the possibility of a pystac Item already having an `href`? Or do we know that will never be the case?

Comment (emmaai) [reply odc/stac/eo3/_eo3converter.py]: that's a good point, I didn't try if either stac item got `href` in it, or loading could work with a `uris` pointing to `odc-yaml`.  Had it been, then there is no need to get url from stac path. I'll try it now.

**Comment by emmaai** [comment]
Sorry I should have read more into the code and understand what's in stac items before making the change.
it clearly states here https://github.com/opendatacube/odc-stac/blob/407db82799aaf823a4f5c1d65bc89eb0357dd06d/odc/stac/_mdtools.py#L701, that the href of stac would be reserved.

**Review by SpacemanPaul — APPROVED** [review]

**Pull request merged**